File: 3/HeapSort.java
import java.util.Arrays;

class Heapsort {
	public static void sort(int arr[]) {
		buildHeap(arr);	
		
		int size = arr.length;
		
		while(size > 0) {
			swap(0, size-1, arr);
			size--;
			siftDown(0, size, arr);
		}
	}

	private static void buildHeap(int arr[]) {
		for(int i=(arr.length-1)/2; i>=0 ; i--)	{
			siftDown(i, arr.length, arr);
		}
	}
	private static void swap(int i, int j, int arr[]) {
		int temp = arr[i];
		arr[i] = arr[j];
		arr[j] = temp;
	}

	private static void siftDown(int i, int size, int arr[]) {
		while(2*i + 1 < size) {
			int left = 2*i+1;
			int right = 2*i+2;

			int selected = left;
			
			if(right < size) {
				if(arr[right] > arr[left]) {
					selected = right;
				}	
			}
			
			if(arr[i] > arr[selected]) {
				return;
			}
			
			swap(i, selected, arr);
			i = selected;
		}	
	}

	public static void main(String args[]) {
		int arr[] = {8,5,2,9,5,6,3};
		sort(arr);
		System.out.println(Arrays.toString(arr));	
	}
}

File: 3/js/src/Heapsort.ts
function swap(i: number, j: number, list: number[]) {
	[list[i], list[j]] = [list[j], list[i]];
}

function siftDown(i: number, list: number[], length: number) {
	while(2*i+1 < length) {
		const left = 2*i + 1;
		const right = 2*i + 2;

		let max = left;

		if(right < length && list[right] > list[left]) {
			max = right;
		}

		if(list[max] < list[i]) {
			return;
		}

		swap(max, i, list);
		i = max;
	}
}

function buildMaxHeap(list: number[]) {
	for(let i=Math.floor(list.length/2)-1; i >= 0 ; i--) {
		siftDown(i, list, list.length);
	}
}

function sort(list: number[]) {
	buildMaxHeap(list);

	for(let i=list.length-1; i>0; i--) {
		swap(0, i, list);
		siftDown(0, list, i);
	}
}

(function main() {
	const list = [8,5,2,9,5,6,3];
	sort(list);
	console.log(list);
})();
